enma/man/enma.1
.Dd Apr 03, 2009
.Dt ENMA 1
.Sh NAME
.Nm enma
.Nd A milter program for domain authentication technologies
.Sh SYNOPSIS
.Nm
.Op Fl cvh
.Op Fl o Ar key=value
.Sh DESCRIPTION
.Nm
is a program of domain authentication technologies. It authenticates
message senders with SPF, Sender ID, DKIM and/or DKIM ADSP and inserts
the Authentication-Results: field with authentication results.
.Pp
Configurations have order of preference as follows:
.Pp
.Bl -enum -offset indent -compact
.It
The command line options
.It
The configuration file
.It
The default values
.El
.Sh OPTIONS
.Bl -tag -width indent
.It Fl h
Displays available options.
.It Fl v
Verbose mode.
.It Fl c Ar filename
Specifies a configuration file.
.It Fl o Ar key=value
Specifies configuration below. 
.El
.Sh CONFIGURATION FILE
Each line is in the format of "key:value". Spaces between "key:" and
"value" are ignored. Lines begging with '#' and empty lines are
ignored. If there is no configuration of a given "key:", its default
value is used.  Absolute paths must be used when specified.
.Bl -tag -width indent
.It milter.verbose (-v)
If true, log is recorded in detail. (Default value: false)
.It milter.conffile (-c)
Specifies the path to the configration file. If ommited, no
configuration file is read and the default values are used.  The -c
option should be used usually instead. (Default value: no value)
.It milter.user
Specifies a user name. If not specified, the user name of execution is
used. This is typically used to step down the user authority of enma
executed in a start-up script.  (Default value: no value)
.It milter.pidfile
Specify the path to the PID file. If milter.user is specified, the PID
file is written as the user. So, appropriate permission should be
given to the directory. (Default value: /var/run/enma/enma.pid)
.It milter.chdir
Specified a working directory after execution. If ommited, nothing
happens. (Default value: no value)
.It milter.socket
Specifies socket type of milter communication from MTA.  Choose one of
the followings:
.Bl -enum -offset indent -compact
.It
inet:<port number>@<IP adddress or hostname>
.It
unix:<a path to UNIX domain socket>
.El
For more information, refer to "cf/README" contained in the Sendmail
package. (Default value: inet:10025@127.0.0.1)
.It milter.timeout
Specifies timeout of milter communication with MTA in seconds.
(Default value: 7210)
.It milter.loglevel
Specifies log level of libmilter (not enma). This should be used for
debugging purpose.  For more information, please refer to
"libmilter/docs/smfi_setdbg.html" in the Sendmail package. (Default
value: 0)
.It milter.sendmail813
If the version of Sendmail is 8.13 or earlier, specify "true",
Otherwise, specify "false". (Default value: false)
.It milter.postfix
If true, MTA is Postfix. If false, MTA is Sendmail. (Default value:
false)
.It syslog.ident
Specified an identifier labeled to syslog messages. (Default value:
enma)
.It syslog.facility
Specifies facility of syslog. (Default value: local4)
.It syslog.logmask
Specifies mask of syslog.  Messages which level is over this value are
printed to syslog. Usually "info" should be specified. (Default
value: info)
.It common.exclusion_addresses
Specifies IP address ranges which are exception of domain
authentication. If the source IP address of the peer matches the
ranges, domain authentication process is omitted. Multiple ranges can
be enumerated with the comma separator. (Default value: 127.0.0.1,::1)
.It spf.auth
If true, SPF authentication is processed.  (Default value: true)
.It spf.explog
If true, the "exp" modifier is evaluated in the case where SPF
authentication result is "hardfail".  For more information about the
"exp" modifier, refer to Section 6.2 of RFC4408.  (Default value:
true)
.It sidf.auth
If true, Sender ID authentication is processed. (Default value: true)
.It sidf.explog
If true, the "exp" modifier is evaluated in the case where Sender ID
authentication result is "hardfail".  For more information about the
"exp" modifier, refer to Section 6.2 of RFC4408.  (Default value:
true)
.It dkim.auth
If true, DKIM authentication is processed. (Default value: true)
.It dkim.signheader_limit
Specifies a limit number of the DKIM-Signature: fields.  They are
evaluated from the top, and are ignored if the number reaches the
limit.  (Default value: 10)
.It dkim.accept_expired_signature
If true, expired DKIM signatures are treated as valid. This value
should be false in normal case. (Default value: false)
.It dkimadsp.auth
If turu, DKIM ADSP check is processed. (Default value: true)
.It authresult.identifier
Specifies the hostname to identity the Authentication-Results:
field. If the Authentication-Results: field which has the same
identifier exists, the entire field is removed. Also, this identifier
is used when the Authentication-Results: field is inserted to record
authentication result.  (Default value: localhost)
.El
.Sh LOG
Log is recored to syslog. facility and mask of syslog are specified
with "syslog.facility" and "syslog.logmask", respectively.
.Sh EXAMPLE OF LOG
The case where authentication process was normally done:
.Bd -literal
[m75AKEOh009630] [SPF-auth] ipaddr=192.168.1.1, eval=smtp.mailfrom, helo=example.jp, envfrom=<user@example.jp>, score=pass
[m75AKEOh009630] [SIDF-auth] ipaddr=192.168.1.1, header.From=user@example.jp, score=pass
[m75AKEOh009630] [DKIM-auth] header.i=user@example.jp, score=pass
[m75AKEOh009630] [DKIM-ADSP-auth] header.from=user@example.jp, score=pass
.Ed
.Pp
The case where authentication process was skipped because of lack of
authentication information:
.Bd -literal
[m75AKEOh009630] [SPF-auth] score=permerror
[m75AKEOh009630] [SIDF-auth] score=permerror
[m75AKEOh009630] [DKIM-auth] score=permerror
[m75AKEOh009630] [DKIM-ADSP-auth] score=permerror
.Ed
.Pp
The case where no signature exists:
.Bd -literal
[m75AKEOh009630] [DKIM-auth] score=none
.Ed
.Sh DESCRIPTION OF LOG ITEM
.Em Common
.Bl -tag -width indent
.It [m75AKEOh009630]
Queue ID to identify SMTP transaction labeled by MTA
.It [SIDF-auth], [SPF-auth], [DKIM-auth], [DKIM-ADSP-auth]
These indicates Sender ID, SPF, DKIM, DKIM ADSP, respectively.
.It ipaddr=192.168.1.1
The IP address of the sender
.It score=pass
Score of authentication result
.El
.Em SPF
.Bl -tag -width indent
.It eval=smtp.mailfrom
Which authentication information was used, either MAIL FROM or
EHLO/HELO.
.It helo=example.jp
EHLO/HELO provided by the sender
.It envfrom=<user@example.jp>
MAIL FROM provided by the sender
.El
.Em Sender ID
.Bl -tag -width indent
.It header.From=user@example.jp
The field name and the mail address used for authentication.
.El
.Em DKIM
.Bl -tag -width indent
.It header.i=user@example.jp
The mail address or domain of a signer.
.El
.Em DKIM ADSP
.Bl -tag -width indent
.It header.From=user@example.jp
The field name and the mail address used for authentication.
.El
.Sh SCORE
See RFC5451.


ChangeLog
2012-01-31  SUZUKI Takahiko  <takahiko@iij.ad.jp>

    * 1.2.0
    * [DKIM] fix mis-canonicalization of empty message body in "relax" mode

2011-12-05  SUZUKI Takahiko  <takahiko@iij.ad.jp>

    * 1.1.992
    * support platforms which doesn't have EAI_NODATA any more
    * adjust configure.ac not to link with librt when ldns is chosen as a resolver

2011-11-28  SUZUKI Takahiko  <takahiko@iij.ad.jp>

    * 1.1.991
    * switch default resolver from libbind to ldns
    * [SPF/SIDF] fix memory allocation error at expanding "p" macro when <ip> has no validated domain names
    * [SPF/SIDF] fix misevaluation on domain names longer than 63 characters (thanks to KODAMA Daisuke, NEC BIGLOBE)
    * [DKIM] catch up the updates from RFC4871, RFC5672 to RFC6376
    * [DKIM] change result from "none" to "permerror" of message without any Author header
    * [DKIM] fix mis-canonicalization of header field name includes WSP in "relax" mode
    * [DKIM] fix memory allocation error on message with syntax errors in sig-i-tag value

2009-04-03  SUZUKI Takahiko  <takahiko@iij.ad.jp>

    * 1.1.0
    * support DKIM (verification only) and DKIM ADSP
    * support exclusion address list to avoid authentication

2008-08-28  SUZUKI Takahiko  <takahiko@iij.ad.jp>

    * 1.0.0
    * public release

2008-08-26  SUZUKI Takahiko  <takahiko@iij.ad.jp>

    * 0.9.2
    * translate documents into English
    * fix pid file permission
    * fix sysconf(_SC_GETPW_R_SIZE_MAX) problem on FreeBSD

2008-08-22  SUZUKI Takahiko  <takahiko@iij.ad.jp>

    * 0.9.1
    * improve build script
    * add man pages

2008-08-19  SUZUKI Takahiko  <takahiko@iij.ad.jp>

    * 0.9.0
    * internal beta release

$Id: ChangeLog 1477 2012-01-30 07:54:30Z takahiko $


enma/src/enma.c
/*
 * Copyright (c) 2008-2011 Internet Initiative Japan Inc. All rights reserved.
 *
 * The terms and conditions of the accompanying program
 * shall be provided separately by Internet Initiative Japan Inc.
 * Any use, reproduction or distribution of the program are permitted
 * provided that you agree to be bound to such terms and conditions.
 *
 * $Id: enma.c 1371 2011-11-07 03:18:02Z takahiko $
 */

#include "rcsid.h"
RCSID("$Id: enma.c 1371 2011-11-07 03:18:02Z takahiko $");

#include <stdio.h>
#include <assert.h>
#include <sysexits.h>
#include <stdlib.h>
#include <syslog.h>

#include <openssl/err.h>
#include <openssl/evp.h>
#include <openssl/crypto.h>
#include <libmilter/mfapi.h>

#include "loghandler.h"
#include "sidf.h"
#include "dkim.h"

#include "cryptomutex.h"
#include "consolehandler.h"
#include "enma_config.h"
#include "enma_mfi.h"
#include "daemonize.h"
#include "enma.h"


/* definition of global variable */
SidfPolicy *g_sidf_policy = NULL;   // sidf policy variable
DkimVerificationPolicy *g_dkim_vpolicy = NULL;  // dkim verify policy variable
EnmaConfig *g_enma_config = NULL;   // configuration variable of ENMA


/**
 * usage of ENMA
 *
 * @param out
 */
static void
enma_usage(FILE *out)
{
    fprintf(out, "Usage:\n");
    fprintf(out, "\tenma [options] [conffile]\n");
    fflush(out);
}


/**
 * initialize configuration
 *
 * @param argc
 * @param argv
 * @return
 */
static int
config_init(int argc, char **argv)
{
    g_enma_config = EnmaConfig_new();
    if (NULL == g_enma_config) {
        return EX_OSERR;
    }

    if (!EnmaConfig_setConfig(g_enma_config, argc, argv)) {
        EnmaConfig_free(g_enma_config);
        enma_usage(stderr);
        return EX_USAGE;
    }

    return 0;
}


/**
 * initialize SIDF policy
 *
 * @param enma_config
 * @return
 */
static SidfPolicy *
sidf_init(const EnmaConfig *enma_config)
{
    SidfPolicy *sidf_policy = SidfPolicy_new();
    if (NULL == sidf_policy) {
        return NULL;
    }
    SidfPolicy_setSpfRRLookup(sidf_policy, false);
    SidfPolicy_setExplanationLookup(sidf_policy, false);
    SidfPolicy_setLogger(sidf_policy, LogHandler_syslogWithPrefix);

    if (SIDF_STAT_OK !=
        SidfPolicy_setCheckingDomain(sidf_policy, enma_config->authresult_identifier)) {
        return NULL;
    }

    return sidf_policy;
}

/**
 * initialize DKIM policy
 *
 * @param *enma_config
 * @return
 */
static DkimVerificationPolicy *
dkim_init(const EnmaConfig *enma_config)
{
    DkimStatus set_stat;

    DkimVerificationPolicy *dkim_vpolicy = DkimVerificationPolicy_new();
    if (NULL == dkim_vpolicy) {
        return NULL;
    }   // end if

    set_stat = DkimVerificationPolicy_setAuthorPriority(dkim_vpolicy, "From", ":");
    if (DSTAT_OK != set_stat) {
        return NULL;
    }   // end if

    DkimVerificationPolicy_setSignHeaderLimit(dkim_vpolicy, enma_config->dkim_signheader_limit);
    DkimVerificationPolicy_acceptExpiredSignature(dkim_vpolicy,
                                                  enma_config->dkim_accept_expired_signature);
    DkimVerificationPolicy_getRfc4871Compatible(dkim_vpolicy, enma_config->dkim_rfc4871_compatible);
    DkimVerificationPolicy_supposeLeadingHeaderValueSpace(dkim_vpolicy,
                                                          enma_config->milter_sendmail813);
    DkimVerificationPolicy_setLogger(dkim_vpolicy, LogHandler_syslogWithPrefix);
    return dkim_vpolicy;
}


/**
 * main function
 *
 * @param argc
 * @param argv
 * @return
 */
int
main(int argc, char **argv)
{
    int result = 0;
    // initialize configuration
    if (0 != (result = config_init(argc, argv))) {
        ConsoleError("enma starting up failed: error=config_init failed");
        exit(result);
    }
    // initialize log handler
    openlog(g_enma_config->syslog_ident, LOG_PID | LOG_NDELAY, g_enma_config->syslog_facility);
    setlogmask(LOG_UPTO(g_enma_config->syslog_logmask));
    LogHandler_init();

    // initialize OpenSSL
    ERR_load_crypto_strings();
    Crypto_mutex_init();

    // initialize SIDF Policy
    if (NULL == (g_sidf_policy = sidf_init(g_enma_config))) {
        ConsoleError("enma starting up failed: error=sidf_init failed");
        exit(EX_OSERR);
    }
    // initialize DKIM Policy
    if (NULL == (g_dkim_vpolicy = dkim_init(g_enma_config))) {
        ConsoleError("enma starting up failed: error=dkim_init failed");
        exit(EX_OSERR);
    }
    // initialize milter
    if (!EnmaMfi_init
        (g_enma_config->milter_socket, g_enma_config->milter_timeout,
         g_enma_config->milter_loglevel)) {
        ConsoleError("enma starting up failed: error=EnmaMfi_init failed");
        exit(EX_OSERR);
    }
    // daemonize
    if (!daemonize_init
        (g_enma_config->milter_user, g_enma_config->milter_chdir, g_enma_config->milter_pidfile)) {
        ConsoleError("enma starting up failed: error=daemonize_init failed");
        LogError("enma starting up failed: error=daemonize_init failed");
        exit(EX_OSERR);
    }

    LogInfo("enma starting up");
    int smfi_return_val = smfi_main();
    LogInfo("enma shutting down: result=%d", smfi_return_val);

    if (!daemonize_finally(g_enma_config->milter_pidfile)) {
        LogError("daemonize_finally failed");
        exit(EX_OSERR);
    }

    SidfPolicy_free(g_sidf_policy);
    DkimVerificationPolicy_free(g_dkim_vpolicy);
    EnmaConfig_free(g_enma_config);

    // OpenSSL cleanup
    Crypto_mutex_cleanup();
    ERR_free_strings();
    EVP_cleanup();
    CRYPTO_cleanup_all_ex_data();

    LogHandler_cleanup();
    closelog();

    return smfi_return_val;
}
